llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,245 @@
+    return std::optional(BundledSCEV(S0, 0));
+
+  if (S1)
+    return std::optional(BundledSCEV(S1, 1));
+
+  LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: Could not extract "
+                       "SCEVAddRecExpr.\nEncountered no ADDRec "
+                       "expressions in the Loop Compare Instruction.");
+  return std::nullopt;
+}
+
+/// Return the Branch Compare Instruction of CurrentLoop if the Loop is well
+/// formed and this pass can process the Predicate
+ICmpInst *getLoopCmpInst(Loop &CurrentLoop) {
+
+  /// Check that the loop has a single Exiting Block. If the CurrentLoop
+  /// contains multiple Exiting Block, ExitBB will be a nullptr
+  auto *ExitBB = CurrentLoop.getExitingBlock();
+  if (!ExitBB)
+    return nullptr;
+
+  BranchInst *BI = dyn_cast<BranchInst>(ExitBB->getTerminator());
+  if (!BI)
+    return nullptr;
+
+  ICmpInst *LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return nullptr;
+
+  LLVM_DEBUG(dbgs() << "Condition Found: " << *LoopCmpInstr << "\n");
+
+  // Do not process equal and not-equal Compare Instructions, since they do not
+  // work well with the currently used assumption-construction-method of reusing
+  // the ICmpInst Predicate
+  if (LoopCmpInstr->getPredicate() == CmpInst::ICMP_EQ ||
+      LoopCmpInstr->getPredicate() == CmpInst::ICMP_NE) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: EQ and NE "
+                         "conditions are currently not supported!");
+    return nullptr;
+  }
+  return LoopCmpInstr;
+}
+
+// Insert a Minimum Iteration Count Assumption into the Loop Header
+void insertMinIterAssumption(ICmpInst &LoopCmpInstr, BasicBlock &LoopHeader,
+                             ScalarEvolution &SE,
+                             const BundledSCEV &BunldedSCEV, int MinIterCount,
+                             AssumptionCache &AC) {
+  Instruction *HeaderTerminator = LoopHeader.getTerminator();
+
+  // LoopRotate uses SimplifyQuery to determine, if a Branch is conditional or
+  // not. SimplifyQuery can only take an Assumption into account, if it is
+  // before the to-be-evaluated Compare Instruction. Thus assumptions have to be
+  // inserted before the LoopCmpInstr.
+  IRBuilder<> Builder(dyn_cast<Instruction>(&LoopCmpInstr));
+
+  SCEVExpander Expander(SE, LoopHeader.getModule()->getDataLayout(),
+                        "expanded");
+
+  // Get Minimum Iteration Value Expression
+  const SCEV *MinIterSCEV = BunldedSCEV.AddRecExpr->evaluateAtIteration(
+      SE.getConstant(APInt(32, MinIterCount)), SE);
+
+  // If the MinIterSCEV is not a constant, but based on an expression,
+  // the no signed/unsigned overflow flags have to be added, so that the
+  // expression can be simplified in later passes
+  SCEVCouldNotCompute::NoWrapFlags NWF = BunldedSCEV.AddRecExpr->getNoWrapFlags(
+      static_cast<SCEVCouldNotCompute::NoWrapFlags>(SCEV::FlagNUW |
+                                                    SCEV::FlagNSW));
```

Comment:
Nit: Please have something like `getNoWrapFlags(/*Mask=*/...)` otherwise it's not really clear what that parameter represents.

Nit2: Is the static cast really needed? Maybe replace with a constructor-style cast like `SCEV::NoWrapFlags(SCEV::FlagNUW | SCEV::FlagNSW)`?